### PR TITLE
cmd/testwrapper: only retry flaky failed tests

### PR DESCRIPTION
**Pull request by maisem**
Redo the testwrapper to track and only retry flaky tests instead
of retrying the entire pkg. It also fails early if a non-flaky test fails.

This also makes it so that the go test caches are used.

Fixes https://github.com/tailscale/tailscale/issues/8438

### CODE REVIEW

**Comment by bradfitz** [comment]
This changed into an entirely different thing. It's no longer an `-exec` wrapper, huh?

**Comment by maisem** [comment]
Yeah I tried using the exec wrapper but would have had to produce the same output format as "go test".

File: cmd/testwrapper/testwrapper.go [hunk 1]
```diff
@@ -1,62 +1,187 @@
 // Copyright (c) Tailscale Inc & AUTHORS
 // SPDX-License-Identifier: BSD-3-Clause
 
-// testwrapper is a wrapper for retrying flaky tests, using the -exec flag of
-// 'go test'. Tests that are flaky can use the 'flakytest' subpackage to mark
-// themselves as flaky and be retried on failure.
+// testwrapper is a wrapper for retrying flaky tests, it is meant to be called
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:4]: either period or semicolon

File: cmd/testwrapper/testwrapper.go [hunk 2]
```diff
@@ -1,62 +1,187 @@
...
+// in place of 'go test'. It assumes the first argument is the package to test,
+// and the rest of the arguments are passed to 'go test'.
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:6]: so you can't pass args to go test itself like "go test -v" ?

File: cmd/testwrapper/testwrapper.go [hunk 3]
```diff
@@ -1,62 +1,187 @@
...
 
-func main() {
-	ctx := context.Background()
-	debug := os.Getenv("TS_TESTWRAPPER_DEBUG") != ""
+const maxAttempts = 3
+
+type testAttempt struct {
+	name          testName
+	outcome       string // "pass", "fail", "skip"
+	logs          []string
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:32]: like one line each?

Comment (bradfitz) [reply cmd/testwrapper/testwrapper.go:32]: (not a bytes.Buffer or something you append to?)

File: cmd/testwrapper/testwrapper.go [hunk 4]
```diff
@@ -1,62 +1,187 @@
...
+		defer close(done)
+		cmd.Wait()
+	}()
 
-	for i := 1; i <= maxIterations; i++ {
-		if i > 1 {
-			log.Printf("retrying flaky tests (%d of %d)", i, maxIterations)
+	jd := json.NewDecoder(r)
+	resultMap := make(map[testName]*testAttempt)
+	var out []testAttempt
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:78]: using this as a value when it has pointers (slice) in it is a little weird. It's a pointer on the line above.

File: cmd/testwrapper/testwrapper.go [hunk 5]
```diff
@@ -1,62 +1,187 @@
...
+	return out
+}
+
+func main() {
+	ctx := context.Background()
+	if len(os.Args) < 2 {
+		fmt.Println("no package specified")
+		os.Exit(1)
+	}
+	pkg, otherArgs := os.Args[1], os.Args[2:]
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:135]: is this purely a single pkg or a single pkg pattern (like "./...")?  It can't do multiple packages+patterns like the Go tool itself? It separates its args with `--`

File: cmd/testwrapper/testwrapper.go [hunk 6]
```diff
@@ -1,62 +1,187 @@
...
+		for _, tr := range runTests(ctx, attempt, pt, otherArgs) {
+			if tr.outcome != "fail" {
+				continue
+			}
+			if tr.isMarkedFlaky {
+				toRetry[tr.name.pkg] = append(toRetry[tr.name.pkg], tr.name.name)
+			} else {
+				failed = true
 			}
+			for _, l := range tr.logs {
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:161]: line

File: cmd/testwrapper/testwrapper.go [hunk 7]
```diff
@@ -1,62 +1,187 @@
...
+	isMarkedFlaky bool // set if the test is marked as flaky
+}
+
+type testName struct {
+	pkg  string
+	name string
+}
+
+type packageTests struct {
+	pkg   string
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:42]: is this a package or pattern?

File: cmd/testwrapper/testwrapper.go (hunk 2 above)
Comment (maisem) [reply cmd/testwrapper/testwrapper.go:6]: you can, i'll document this better.

This works `testwrapper ./... -v`.

Comment (bradfitz) [reply cmd/testwrapper/testwrapper.go:6]: can you make it behave like go test instead? including with its `-args` support?

https://pkg.go.dev/cmd/go#hdr-Testing_flags and search for the two "For instance, the command" bits.

Comment (bradfitz) [reply cmd/testwrapper/testwrapper.go:6]: Alternatively, don't even try to match and do something obviously different.

Comment (maisem) [reply cmd/testwrapper/testwrapper.go]: PTAL

File: cmd/testwrapper/testwrapper.go (hunk 5 above)
Comment (maisem) [reply cmd/testwrapper/testwrapper.go]: it can't do multiple packages/patterns. I didn't try implementing that.

File: cmd/testwrapper/testwrapper.go [hunk 8]
```diff
@@ -1,62 +1,223 @@
 // Copyright (c) Tailscale Inc & AUTHORS
 // SPDX-License-Identifier: BSD-3-Clause
 
-// testwrapper is a wrapper for retrying flaky tests, using the -exec flag of
-// 'go test'. Tests that are flaky can use the 'flakytest' subpackage to mark
-// themselves as flaky and be retried on failure.
+// testwrapper is a wrapper for retrying flaky tests. It is meant to be called
+// in place of 'go test'. It requires the first argument is the pattern to test,
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:5]: Can you reword this to not suggest it's a "go test" drop-in replacement? Just say it's its own thing. Maybe "It's an alternative to 'go test' that re-runs tests as needed. It takes different arguments than go test: it requires ..."

File: cmd/testwrapper/testwrapper.go [hunk 9]
```diff
@@ -1,62 +1,223 @@
...
+
+type testAttempt struct {
+	name          testName
+	outcome       string // "pass", "fail", "skip"
+	logs          bytes.Buffer
+	isMarkedFlaky bool // set if the test is marked as flaky
+}
+
+type testName struct {
+	pkg  string
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:39]: example strings on all these fields?

```
pkg string // "tailscale.com/types/key"
name string // "TestFoo"

...

pattern string // "./..."
tests string // ??
```

File: cmd/testwrapper/testwrapper.go [hunk 10]
```diff
@@ -1,62 +1,223 @@
...
+			}
+		}
+		if failed {
 			os.Exit(1)
 		}
-
-		if code := exitErr.ExitCode(); code != retryStatus {
-			if debug {
-				log.Printf("code (%d) != retryStatus (%d)", code, retryStatus)
+		for pkg, tests := range toRetry {
```
Comment (bradfitz) [inline cmd/testwrapper/testwrapper.go:204]: can you do this in a deterministic order?

**Review by bradfitz — APPROVED** [review]
